### ISSUE TITLE
finalizer of `Bracket. guarantee` is not executed upon `fiber.cancel`

### ISSUE THREAD

**Issue by t3hnar** [typelevel/cats-effect#461]
I found something weird and cannot explain so far.
Looks like only last guarantee is taken into account during cancellation for abstract `F[_]`

```scala
          .guarantee(ref1.set(true))
          .guarantee(ref2.set(true))
```
However that works fine when called directly against `IO`
Also looks like @kubukoz can see variant results for both `IO` and `F[_]` examples.

Here is reproducible snippet

```scala
import cats.effect._
import cats.effect.concurrent.Ref
import cats.effect.implicits._

import scala.concurrent.duration._

object Example extends IOApp {

  def run(args: List[String]): IO[ExitCode] = {
    //    runF[IO]
    runIO
  }


  def runIO: IO[ExitCode] = {
    for {
      ref1 <- Ref.of[IO, Boolean](false)
      ref2 <- Ref.of[IO, Boolean](false)
      fiber <- start[IO](ref1, ref2)
      _ <- fiber.cancel
      _ <- IO.sleep(100.millis)
      r1 <- ref1.get
      r2 <- ref2.get
    } yield {
      println(s"r1: $r1, r2: $r2")
      ExitCode.Success
    }
  }


  def runF[F[_] : Concurrent](implicit timer: Timer[F]) = {
    import cats.implicits._
    for {
      ref1 <- Ref.of[F, Boolean](false)
      ref2 <- Ref.of[F, Boolean](false)
      fiber <- start[F](ref1, ref2)
      _ <- fiber.cancel
      _ <- timer.sleep(100.millis)
      r1 <- ref1.get
      r2 <- ref2.get
    } yield {
      println(s"r1: $r1, r2: $r2")
      ExitCode.Success
    }
  }

  def start[F[_] : Concurrent](ref1: Ref[F, Boolean], ref2: Ref[F, Boolean]) = {
    Concurrent[F].start {
      Async[F].never[Unit]
        .guarantee(ref1.set(true))
        .guarantee(ref2.set(true))
    }
  }
}
```

**Comment by alexandru** [comment]
When you’re starting a fiber and then immediately canceling it, you do not have the guarantee that the computation will actually start and thus have no guarantee that a finalizer will be installed.

If you’ll take a look at all of our tests that observe cancellation, we use latches for waiting on the computation to be started, before canceling the fiber.

**Comment by t3hnar** [comment]
@alexandru the issue I'm trying to show - that there are *two finalizere` and one is executed but another one - not.
So basically `fiber` is started.

**Comment by alexandru** [comment]
In other words your reproducing snippet isn’t correct because its result depends on a race condition, i.e. the fiber has to execute faster than the main thread doing the cancellation. Sometimes this can happen, but most of the times it won’t, due to the fiber execution involving a thread fork, which is expensive.

**Comment by t3hnar** [comment]
@alexandru I can agree that snippet is not correct from that perspective. Should I fix it ?

Anyhow, was I able to convince you about some strange issue taking place there? :)

**Comment by alexandru** [comment]
The second finalizer follows the same rules @t3hnar. Due to the auto-cancellation model you can end up with only the first finalizer executing, because the second is never evaluated.

In other words there is nothing on your sample that guarantees _atomic execution_ of the resulting computation.

**Comment by t3hnar** [comment]
> what you need is a full bracket.

In my case I want to `acquire` resource on main thread, to make sure we can `acquire` it in general.
And the pass it to forked thread, which will `release` when done.
Hence full `bracket` does not suit here.

Looking into code I cannot see why `guarantee` is a bad choice for my case.

**Comment by alexandru** [comment]
> Do you happen to know why I'm getting those different for IO and F[_] ?

I think that’s just non-determinism. Run them enough times and you should see all combinations, including `(false, false)`.

**Comment by alexandru** [comment]
On `guarantee` being a bad choice, that’s because you won’t get the guarantee that the finalizer will be installed and executed. As you can see here.

You need to acquire the resource via `bracket` and then use its `release` for the finalizer. Note both `use` and `release` can fork another thread.

Unfortunately this is the drawback of the auto-cancellation model. Resource acquisition needs `bracket`.

Monix’s `Task` can work with non-interruptible flatMaps, so you can have an easier time on the use case you’re thinking of, but we have no type class that exposes this yet, so polymorphic code would be broken.

**Comment by t3hnar** [comment]
I changed example and can confirm that for this is non-determinism for both `F[_]` and `IO`

```
import cats.effect._
import cats.effect.concurrent.{Deferred, Ref}
import cats.effect.implicits._
import cats.implicits._

import scala.concurrent.duration._

object Example extends IOApp {

  def run(args: List[String]): IO[ExitCode] = {
    val fas = List.fill(100)(runF[IO]) ++ List.fill(100)(runIO)
    fas.foldLeft(IO.unit)(_ *> _).as(ExitCode.Success)
  }


  def runIO: IO[Unit] = {
    for {
      deferred <- Deferred[IO, Unit]
      ref1 <- Ref.of[IO, Boolean](false)
      ref2 <- Ref.of[IO, Boolean](false)
      fiber <- start[IO](deferred, ref1, ref2)
      _ <- deferred.get
      _ <- fiber.cancel
      _ <- IO.sleep(100.millis)
      r1 <- ref1.get
      r2 <- ref2.get
    } yield {
      println(s"IO r1: $r1, r2: $r2")
    }
  }


  def runF[F[_] : Concurrent](implicit timer: Timer[F]): F[Unit] = {
    for {
      deferred <- Deferred[F, Unit]
      ref1 <- Ref.of[F, Boolean](false)
      ref2 <- Ref.of[F, Boolean](false)
      fiber <- start[F](deferred, ref1, ref2)
      _ <- deferred.get
      _ <- fiber.cancel
      _ <- timer.sleep(100.millis)
      r1 <- ref1.get
      r2 <- ref2.get
    } yield {
      println(s"F[_] r1: $r1, r2: $r2")
    }
  }

  def start[F[_] : Concurrent](
    deferred: Deferred[F, Unit],
    ref1: Ref[F, Boolean],
    ref2: Ref[F, Boolean]) = {

    Concurrent[F].start {
      for {
        _ <- deferred.complete(())
        _ <- Async[F].never[Unit]
          .guarantee(ref1.set(true))
          .guarantee(ref2.set(true))
      } yield {}
    }
  }
}
```

**Comment by t3hnar** [comment]
Here is working example of feature I was looking for 
Basically implementation of forking usage of `Resource` and still making sure that `acquire` did work before we move on with the main thread.

```scala
import cats.effect._
import cats.effect.concurrent.{Deferred, Ref}
import cats.implicits._
import cats.effect.implicits._

object Example extends IOApp {

  def run(args: List[String]): IO[ExitCode] = {
    val fas = List.fill(10000)(runF[IO])
    fas.foldLeft(IO.unit)(_ *> _).as(ExitCode.Success)
  }


  def runF[F[_] : Concurrent]: F[Unit] = {
    for {
      ref <- Ref.of[F, Boolean](false)
      res = Resource.make(().pure[F])(_ => ref.set(true))
      fiber <- startRes(res)(_.pure[F])
      _ <- fiber.cancel
      result <- ref.get
    } yield {
      if (!result) println(s"wrong")
    }
  }

  def startRes[F[_] : Concurrent, A, B](res: Resource[F, A])(use: A => F[B]): F[Fiber[F, B]] = {
    for {
      acquired <- Deferred[F, Option[Throwable]]
      released <- Deferred[F, Unit]
      fiber <- Concurrent[F].start {
        Bracket[F, Throwable].bracket {
          {
            for {
              a <- res.allocated.attempt
              _ <- acquired.complete(a.fold(_.some, _ => none))
            } yield a
          }.rethrow
        } { case (a, _) => use(a)
        } { case (_, release) => release.guarantee(released.complete(())) }
      }
      _ <- acquired.get
    } yield {
      new Fiber[F, B] {

        def cancel = {
          for {
            _ <- fiber.cancel
            _ <- released.get
          } yield {}
        }

        def join = fiber.join
      }
    }
  }
}
```

@alexandru do you have any concerns about this?

**Comment by t3hnar** [comment]
or even simpler case, but with using `allocate` and `release` on different threads.

```scala
def startRes[F[_] : Concurrent, A, B](resource: Resource[F, A])(use: A => F[B]): F[Fiber[F, B]] = {
    for {
      released <- Deferred[F, Unit]
      allocated <- resource.allocated
      (a, release) = allocated
      deferred <- Deferred[F, Unit]
      fiber <- Concurrent[F].start {
        {
          for {
            _ <- deferred.complete(())
            b <- use(a)
          } yield b
        }.guarantee(release.guarantee(released.complete(())))
      }
      _ <- deferred.get
    } yield {
      new Fiber[F, B] {

        def cancel = {
          for {
            _ <- fiber.cancel
            _ <- released.get
          } yield {}
        }

        def join = fiber.join
      }
    }
  }
```
